server/src/main/java/org/opensearch/common/blobstore/transfer/RemoteTransferContainer.java

```diff
@@ -0,0 +1,180 @@
+     * @param offsetRangeInputStreamSupplier A supplier to create OffsetRangeInputStreams
+     */
+    public RemoteTransferContainer(
+        String localFileName,
+        String remoteFileName,
+        long contentLength,
+        boolean failTransferIfFileExists,
+        WritePriority writePriority,
+        OffsetRangeInputStreamSupplier offsetRangeInputStreamSupplier
+    ) {
+        this.localFileName = localFileName;
+        this.remoteFileName = remoteFileName;
+        this.contentLength = contentLength;
+        this.failTransferIfFileExists = failTransferIfFileExists;
+        this.writePriority = writePriority;
+        this.offsetRangeInputStreamSupplier = offsetRangeInputStreamSupplier;
+    }
+
+    /**
+     * @return The {@link  WriteContext} for the current upload
+     */
+    public WriteContext createWriteContext() {
+        return new WriteContext(
+            remoteFileName,
+            this::supplyStreamContext,
+            contentLength,
+            failTransferIfFileExists,
+            writePriority,
+            this::finalizeUpload
+        );
+    }
+
+    /**
+     * @param partSize Part sizes of all parts apart from the last one, which is determined internally
+     * @return The {@link StreamContext} object that will be used by the vendor plugin to retrieve streams
+     */
+    public StreamContext supplyStreamContext(long partSize) {
+        try {
+            return this.openMultipartStreams(partSize);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private StreamContext openMultipartStreams(long partSize) throws IOException {
+        if (inputStreams.get() != null) {
+            throw new IOException("Multi-part streams are already created.");
+        }
+
+        this.partSize = partSize;
+        this.lastPartSize = (contentLength % partSize) != 0 ? contentLength % partSize : partSize;
+        this.numberOfParts = (int) ((contentLength % partSize) == 0 ? contentLength / partSize : (contentLength / partSize) + 1);
+        ResettableCheckedInputStream[] streams = new ResettableCheckedInputStream[numberOfParts];
+        inputStreams.set(streams);
+
+        return new StreamContext(new StreamProvider(getTransferPartStreamSupplier(), partSize, lastPartSize, numberOfParts), numberOfParts);
+    }
+
+    private TransferPartStreamSupplier getTransferPartStreamSupplier() {
+        return ((partNo, size, position) -> {
+            assert inputStreams.get() != null : "expected inputStreams to be initialised";
+            return getMultipartStreamSupplier(partNo, size, position).get();
+        });
+    }
+
+    /**
+     * OffsetRangeInputStreamSupplier is used to get the offset based input streams at runtime
+     */
+    public interface OffsetRangeInputStreamSupplier {
+        OffsetRangeInputStream get(long size, long position) throws IOException;
+    }
+
+    interface LocalStreamSupplier<Stream> {
```

Comment:
can you please help me understand the need of this as well ?

Reply:
It is just used to pass the function used for stream creation to the plugin layer, so that they are called during uploads. We should be able to use the java `Supplier` here. Mostly, these interfaces were added to just have better readability as to what they would be doing, but using existing utilities should be okay.